core/src/jsMain/kotlin/dev/fritz2/remote/http.kt

```diff
@@ -6,229 +6,306 @@ import org.w3c.fetch.*
+    val url: String = "",
+    val method: String = "",
+    val headers: Map<String, String> = emptyMap(),
+    val body: dynamic = undefined,
+    val referrer: String? = undefined,
+    val referrerPolicy: dynamic = undefined,
+    val mode: RequestMode? = undefined,
+    val credentials: RequestCredentials? = undefined,
+    val cache: RequestCache? = undefined,
+    val redirect: RequestRedirect? = undefined,
+    val integrity: String? = undefined,
+    val keepalive: Boolean? = undefined,
+    val reqWindow: Any? = undefined,
+    val interceptors: List<Interceptor> = emptyList()
 ) {
 
     /**
-     * builds a request, sends it to the server, awaits the response (async), creates a flow of it and attaches the defined errorHandler
+     * creates a copy of the [Request].
+     */
+    open fun copy(
+        url: String = this.url,
+        method: String = this.method,
+        headers: Map<String, String> = this.headers,
+        body: dynamic = this.body,
+        referrer: String? = this.referrer,
+        referrerPolicy: dynamic = this.referrerPolicy,
+        mode: RequestMode? = this.mode,
+        credentials: RequestCredentials? = this.credentials,
+        cache: RequestCache? = this.cache,
+        redirect: RequestRedirect? = this.redirect,
+        integrity: String? = this.integrity,
+        keepalive: Boolean? = this.keepalive,
+        reqWindow: Any? = this.reqWindow,
+        interceptors: List<Interceptor> = this.interceptors,
+    ) = Request(
+        url, method, headers, body, referrer, referrerPolicy,
+        mode, credentials, cache, redirect, integrity, keepalive,
+        reqWindow, interceptors
+    )
+
+    /**
+     * executes the HTTP call and sends it to the server, awaits the response (async) and returns a [Response].
+     * When request failed a [FetchException] will be thrown.
      *
-     * @param subUrl function do derive the url (so you can use baseUrl)
-     * @param init an instance of [RequestInit] defining the attributes of the request
-     */
-    private suspend fun execute(subUrl: String, init: RequestInit): Response {
-        val url = buildString {
-            append(baseUrl.trimEnd('/'))
-            if (subUrl.isNotEmpty()) {
-                append("/${subUrl.trimStart('/')}")
-            }
-        }
-        var response = browserWindow.fetch(url, init).await()
-        if (authentication != null) {
-            if (authentication.errorcodesEnforcingAuthentication.contains(response.status)) {
-                authentication.authenticate()
-                val redo = authentication.enrichRequest(this).buildInit(init.method!!)
-                response = browserWindow.fetch(url, redo).await()
-            }
-        }
+     * @throws FetchException when request failed
+     */
+    suspend fun execute(): Response {
+
+        var request = this
+        for(interceptor in interceptors) request = interceptor.enrichRequest(request)
```

Comment:
This approach does apply all interceptors in any case. The interceptor itself has no possibility to abort the upcoming interceptors. I am not sure whether this feature is needed (often), but other frameworks rely on the "chain of responsibility" pattern instead (servelt API for example). It might be possible though, that the latter simply is not needed at client side. But it might be worth thinking about before we finalize this API!

One can quite easily realize this of course by implementing such a chaining mechanism within a single interceptor and then manage all "real" interceptors from within. But that would bypass our API and would feel quite cumbersome.